packages/@lwc/template-compiler/src/parser/expression-complex/validate.ts

```diff
@@ -0,0 +1,152 @@
+        MetaProperty: 'import.meta',
+        NewExpression: 'object instantiation',
+        RegExpLiteral: 'regular expression literals',
+        Super: '`super`',
+        ThisExpression: '`this`',
+        YieldExpression: '`yield`',
+    })
+);
+const STATEMENT_TYPES = new Set([
+    'Statement',
+    'BlockStatement',
+    'BreakStatement',
+    'ContinueStatement',
+    'DebuggerStatement',
+    'DoWhileStatement',
+    'EmptyStatement',
+    'ExpressionStatement',
+    'ForInStatement',
+    'ForStatement',
+    'FunctionDeclaration',
+    'IfStatement',
+    'LabeledStatement',
+    'ReturnStatement',
+    'SwitchStatement',
+    'ThrowStatement',
+    'TryStatement',
+    'VariableDeclaration',
+    'WhileStatement',
+    'WithStatement',
+    'ClassDeclaration',
+    'ExportAllDeclaration',
+    'ExportDefaultDeclaration',
+    'ExportNamedDeclaration',
+    'ForOfStatement',
+    'ImportDeclaration',
+    'DeclareClass',
+    'DeclareFunction',
+    'DeclareInterface',
+    'DeclareModule',
+    'DeclareModuleExports',
+    'DeclareTypeAlias',
+    'DeclareOpaqueType',
+    'DeclareVariable',
+    'DeclareExportDeclaration',
+    'DeclareExportAllDeclaration',
+]);
+const EARLY_STAGE_LANGUAGE_FEATURES = new Set([
+    'BindExpression',
+    'DecimalLiteral',
+    'DoExpression',
+    'RecordExpression',
+    'TupleExpression',
+    'ModuleExpression',
+    'TopicReference',
+    'PipelineTopicExpression',
+    'PipelineBareFunction',
+    'PipelinePrimaryTopicReference',
+]);
+const MUTATION_TYPES = new Set(['AssignmentExpression', 'UpdateExpression']);
+
+function validateArrowFunction(node: t.ArrowFunctionExpression) {
+    invariant(node.body.type !== 'BlockStatement', ParserDiagnostics.INVALID_EXPR_ARROW_FN_BODY);
+    invariant(!node.async, ParserDiagnostics.INVALID_EXPR_ARROW_FN_KIND, ['async']);
+    // This condition should never occur, unless the spec changes. However, it is
+    // permitted by the ESTree representation, so we'll check for it just in case.
+    invariant(!node.generator, ParserDiagnostics.INVALID_EXPR_ARROW_FN_KIND, ['generators']);
+}
+
+function validateUnaryExpression(node: t.UnaryExpression) {
+    invariant(node.operator !== 'delete', ParserDiagnostics.INVALID_EXPR_DELETE_OP);
+}
+
+function validateLiteral(node: t.Literal) {
+    invariant(
+        (node as t.BigIntLiteral).bigint === undefined,
+        ParserDiagnostics.INVALID_EXPR_PROHIBITED_NODE_TYPE,
+        ['BigInts']
+    );
+    invariant(
+        (node as t.RegExpLiteral).regex === undefined,
+        ParserDiagnostics.INVALID_EXPR_PROHIBITED_NODE_TYPE,
+        ['regular expression literals']
+    );
+}
+
+function validateNode(node: BaseNode, _parent: BaseNode, isWithinArrowFn: boolean) {
+    invariant(
+        !node.leadingComments?.length && !node.trailingComments?.length,
+        ParserDiagnostics.INVALID_EXPR_COMMENTS_DISALLOWED
+    );
+    invariant(
+        !STATEMENT_TYPES.has(node.type),
```

Comment:
Instead of blocking statements nodes, I think it is more future-proof to allow list expressions. This way we would have control over which expressions we allow in the template as new expression types are added to the language.

Reply:
I originally created an allowlist for node types, but it was messier to translate the decisions in the design doc this into code. I'll probably revisit, but likely in a follow-up PR.